agent/agent.go

```diff
@@ -403,29 +410,34 @@ func (a *Agent) GetInputPlugin(name string) (telegraf.Input, error) {
 	return nil, fmt.Errorf("could not find input with name: %s", name)
 }
 
+// GetDefaultInputPlugin gets the default InputConfig for a default plugin given its name
+func (a *Agent) GetDefaultInputPlugin(name string) (telegraf.Input, error) {
+	p, exists := inputs.Inputs[name]
+	if exists {
+		return p(), nil
+	}
+	return nil, fmt.Errorf("could not find input with name: %s", name)
+}
+
 // UpdateInputPlugin gets the InputConfig for a plugin given its name
 func (a *Agent) UpdateInputPlugin(name string, config map[string]interface{}) (telegraf.Input, error) {
-
 	for _, input := range a.Config.Inputs {
 		if name == input.Config.Name {
-			// TODO: STOP PLUGIN
 			plugin := input.Input
-			shallowPluginCopy := plugin
 
-			reflectedPlugin := reflect.ValueOf(shallowPluginCopy)
+			a.wg.Add(1)
+
+			a.StopInputPlugin(name)
+
+			reflectedPlugin := reflect.ValueOf(plugin)
 			_, err := updateStructValuesHelper(reflectedPlugin, config)
-			// only update the input if there was no error, indicating
-			// that all fields were updated properly
-			if err == nil {
-				log.Printf("D! updating input, no error!")
-				input.Input = shallowPluginCopy
-			}
 
-			// TODO: START PLUGIN
+			a.RunSingleInput(input.Config, plugin, a.Context)
+			a.wg.Done()
```

Comment:
Doesn't run single input already wait on the wg? Why do we need to wait here? Don't we want to immediately return whether the plugin was able to be updated on not? As it is rn, this function will block until Telegraf ends and will block whatever goroutine called it

Reply:
Hack to deal with the case of 1 plugin. When there is only 1 plugin and we call stop, the only wait group counter gets taken out.

Reply:
Maybe we can add a comment to explain since it is not intuitive